src/ripple/nodestore/impl/Database.cpp

```diff
@@ -312,19 +290,20 @@ Database::threadEntry()
             auto it = read_.lower_bound(readLastHash_);
             if (it == read_.end())
             {
+                // start over from the beginning
                 it = read_.begin();
-                // A generation has completed
-                ++readGen_;
-                readGenCondVar_.notify_all();
             }
             lastHash = it->first;
-            lastSeq = it->second;
+            entry = std::move(it->second);
             read_.erase(it);
             readLastHash_ = lastHash;
         }
 
-        // Perform the read
-        fetchNodeObject(lastHash, lastSeq, FetchType::async);
+        for (auto const& req : entry)
+        {
+            auto obj = fetchNodeObject(lastHash, req.first, FetchType::async);
```

Comment:
How often do we expect to have multiple callbacks in the `entry` collection with the same ledgerSeq? When that happens we make redundant calls to `fetchNodeObject` that we don't need.

Two suggestions to address this:
* Sort the `entry` vector by the first element in the pair, and keep track of the last ledgerIndex used in `fetchNodeObject` and only call that function when `req.first` is different from the last call
* Instead of entry being a vector of pairs, it could be a `unordered_map<std::uint32_t,std::vector<std::function<std::shared_ptr<NodeObject>&)>>>>`

Of course, if we think this is a rare case it might not be worth handling it.

Reply:
@seelabs It is redundant only when the ledger sequence and the node object hash are the same. My guess is that's unlikely but it should be examined closer.

@JoelKatz You are using the same node object hash for every fetch. You may need to make `entry` the same type as `read_` to retain the hash.

Reply:
I think this is sufficiently unlikely to happen and the consequences of it happening (an extra read) so mild that I left the code as simple as possible.

Reply:
Suppose the same node is in two different ledgers, each in a different shard. We may be searching for both nodes simultaneously. It is critical that we neither tell one query that the node is not in the shard if it is and also critical that we not tell one query that the node is in the shard if it isn't. Otherwise, we may build incomplete shards and fail to retrieve from complete ones!